src/main/java/com/iota/iri/service/snapshot/impl/LocalSnapshotServiceImpl.java

```diff
@@ -0,0 +1,459 @@
+     * {@link SnapshotProvider}.
+     *
+     * It first writes the files to the disk and then updates the two {@link Snapshot}s accordingly.
+     *
+     * @param snapshotProvider data provider for the {@link Snapshot}s that are relevant for the node
+     * @param newSnapshot Snapshot that shall be persisted
+     * @param config important snapshot related configuration parameters
+     * @throws SnapshotException if anything goes wrong while persisting the snapshot
+     */
+    private void persistLocalSnapshot(SnapshotProvider snapshotProvider, Snapshot newSnapshot, SnapshotConfig config)
+            throws SnapshotException {
+
+        newSnapshot.writeToDisk(config.getLocalSnapshotsBasePath());
+
+        snapshotProvider.getInitialSnapshot().lockWrite();
+        snapshotProvider.getLatestSnapshot().lockWrite();
+
+        snapshotProvider.getInitialSnapshot().update(newSnapshot);
+
+        snapshotProvider.getLatestSnapshot().setInitialHash(newSnapshot.getHash());
+        snapshotProvider.getLatestSnapshot().setInitialIndex(newSnapshot.getIndex());
+        snapshotProvider.getLatestSnapshot().setInitialTimestamp(newSnapshot.getTimestamp());
+
+        snapshotProvider.getInitialSnapshot().unlockWrite();
+        snapshotProvider.getLatestSnapshot().unlockWrite();
+    }
+
+    /**
+     * This method determines if a transaction is orphaned.
+     *
+     * Since there is no hard definition for when a transaction can be considered to be orphaned, we define orphaned in
+     * relation to a referenceTransaction. If the transaction or any of its direct or indirect approvers saw a
+     * transaction being attached to it, that arrived after our reference transaction, we consider it "not orphaned".
+     *
+     * Since we currently use milestones as reference transactions that are sufficiently old, this definition in fact is
+     * a relatively safe way to determine if a subtangle "above" a transaction got orphaned.
+     *
+     * @param tangle Tangle object which acts as a database interface
+     * @param transaction transaction that shall be checked
+     * @param referenceTransaction transaction that acts as a judge to the other transaction
+     * @param processedTransactions transactions that were visited already while trying to determine the orphaned status
+     * @return true if the transaction got orphaned and false otherwise
+     * @throws SnapshotException if anything goes wrong while determining the orphaned status
+     */
+    private boolean isOrphaned(Tangle tangle, TransactionViewModel transaction,
+            TransactionViewModel referenceTransaction, Set<Hash> processedTransactions) throws SnapshotException {
+
+        long arrivalTime = transaction.getArrivalTime() / 1000L + ORPHANED_TRANSACTION_GRACE_TIME;
+        if (arrivalTime > referenceTransaction.getTimestamp()) {
+            return false;
+        }
+
+        AtomicBoolean nonOrphanedTransactionFound = new AtomicBoolean(false);
```

Comment:
You can not modify variables inside a lambda. Thats why we have to use a mutable object for this flag.

Reply:
renaming it to orphanedTransactionFound doesnt work because we are looking for non orphaned ones.

If i find a aingle orphaned transaction it doesnt mean that this transaction doesnt maybe have another non orphaned one which would in return make this transaction non-orphaned.